Revisiting syscall

agge3 opened the pull request.

agge3 committed:
added design.md and scret.lua

agge3 committed:
updated design.md

agge3 committed:
added additional thoughts

agge3 committed:
A lot of mess right now, because a lot of things need to come over from original
Working towards a working class scarg
Brought the pieces from process_args() over
Separated out syscall:addargs() in class syscall
Seems close to working, but fine details to flesh out
Have not arrived at a testing point yet, and does not run

agge3 committed:
Separated states in class syscall
Continued to work on class scarg
Separated syscall add precedure in class FreeBSDSyscall into respective states
Ready to begin testing, so preparing for testing branch

agge3 committed:
Testing the library to get to a working point
Works as before
States need work and are not called

agge3 committed:
Merge pull request #1 from agge3/test

Merging work done from test

agge3 committed:
Explicit states in syscall:add()
Dumping tables with local DataDumper
Errors with parsing args. Working on

agge3 committed:
Everything's working as it should
Not final implementation
Still need to attend to default config
Good starting point

agge3 committed:
Merge pull request #2 from agge3/test

Merging testing branch

agge3 committed:
Added class scret
class syscall uses scret in addfunc()
Cleaned up other things

agge3 committed:
Merge pull request #3 from agge3/class-scret

Added class scret

agge3 committed:
Added files for all modules
Added license and author(s) credit
Preparing to sort out config more to have a more final solution

agge3 committed:
Dates in license were wrong
Fixed

agge3 committed:
Continuing to sort out what needs to go where
Added necessary global functions to config
Moved functions from scarg to config
Added config data in places where it's needed
Removed config data in places where it's not needed

agge3 committed:
Merge pull request #4 from agge3/config

Merging changes from config branch

agge3 committed:
Added util.generated_tag(str)
Common procedure for generated tag
Takes different string per module, for each file
Updated all modules that use

agge3 committed:
Merge pull request #5 from agge3/generated

Added util.generated_tag(str)

agge3 committed:
Finalized util.generated_tag()
Handles different comments as optional parameter
Handles multi-line titles
Documented function
Tested in lua REPL

agge3 committed:
Merge pull request #6 from agge3/generated

Finalized util.generated_tag()

agge3 committed:
Sorted out util and config properly
Less subject to change
Added config.merge_global()
Calling config.merge_global() instead of naked if condition

agge3 committed:
Merge pull request #7 from agge3/config

Sorted out util and config properly

agge3 committed:
Added class bsdio

To wrapper standard lua assert(fh:write()) as a simpler call
To package common FreeBSD syscall io procedures as a reusable interface
To (maybe) have different behavior based on environment variables
Did not remove generated_tag() from util, or change call in modules
To avoid breakage until confirmed to be a good approach
Sponsored by:    Google (GSoC 2024)

agge3 committed:
scarg:add() no longer creates nested table

Changed to scarg:append(tbl)
Added scarg:pad(tbl)
Adding (and padding) each argument line to syscall argument table
is now properly addressed
Tested with local DataDumper, tables look good
Sponsored by:    Google (GSoC 2024)

agge3 committed:
class scret performs its functions isolated
Can be extended to add more
Revisited implementation of syscall:add()
Removed syscall:is_added(), doesn't fit correctly
Revised implementation of syscall:add() performs the function well
Other small tweaks

agge3 committed:
Fixed is_added()

bsdimp commented:
Why'd you close this? Do you have a new one?

Last week was insane for me due to issues at $WORK, but this week was looking fairly light so I can review this.

agge3 commented:
Oh? I didn't know you opened the PR, I thought I may have accidentally.
I wasn't sure if you wanted me to PR.
In terms of stable changes, this PR is fine.
It's one commit behind and that commit was a one line semantic.
I'll reopen!
I have a branch, separating-modules, but it's super WIP.
Wasn't for that reason though, I know you guys are busy!

bsdimp commented:
You'd opened it originally :)...  I'm happy to review here, though.
And please do fix the style issues. I write lua in, at best, a ham-fisted way, ignorant of style.lua.

agge3 commented:
Oopss! Well, sorry about that..! That works, thank you.
Ah, that explains why the style was inconsistent.. haha. I'll follow style.lua!
I'm going to move towards having a cleaner diff for separating-modules work, and also include Kyle's recommendations.

agge3 commented:
Following up on that, I cleaned-up a lot of the WIP from separating-modules and merged into the main branch of my fork. So, there can be a good milestone of where I'm at.
Also, I've corrected much of the style to style.lua(9), along with commenting as per Kyle's recommendation.
So, there's a lot of new work!
I'm closer to an answer on init_sysent and closer to an answer on what data/procedures need to be where at and what times.
Let me know, when you'd like the see the new work -- I can PR or just keep it on my fork, whatever works.
I'm expecting that once everything right now has a concrete answer (which init_sysent pretty much forces), the rest of the modules will fall in place nicely.

bsdimp commented on design.md:
```diff
@@ -0,0 +1,128 @@
+probably do what it needs to do by default, and pass inclusion/exclusion args
+    e.g., ./makesyscalls --include=init_sysent,syscall  # only does init_sysent.h and syscall.c
+          ./makesyscalls --exclude=systrace_args        # does everything except systrace_args.h
+    also a great place to provide man page and further online documentation
```
I'd go farther and say `that` makesyscalls should 100% disappear.
It was one big monolithic program because it was just easier back in the day to parse once and output many things. Now, I don't think that makes much sense and you'd waste a lot of time on include/exclude that I think we don't really need anymore.

kevans91 replied:
IMO splitting generation of the various files out into their own front-end and targets off of `sysent` would be a good test of the lib'ified model, making sure it's as ergonomic as we'd like for these different exact needs to use.

agge3 replied:
This is somewhat of a crossroads. I'm leaning more towards splitting it off for the testing aspect, but I want to stay focused on what you guys need, so I'll put a pin in that for now.

bsdimp commented on makesyscalls.lua:
```diff
@@ -37,18 +35,21 @@ local unistd = require("posix.unistd")
 	sysproto_h = "_SYS_SYSPROTO_H_",
 	syshdr = "../sys/syscall.h",
-	sysmk = "../sys/syscall.mk",
+	sysmk = "/dev/null",
```
This is an example of something we can definitely generate on the fly.
The syscall.mk file just has one line that has a list of MIASM files. We can instead just have libsys generate it on the fly.
As we move things out of this monolith into the build system that either creates the file when syscalls.master has changed, or whatever, we can also, imho, just remove the code from here until there's nothing left.

bsdimp replied:
I think the same is true of libsys_h, libsysmap, syscalls.c easily fall under this. sysproto.h may as well. I fear that we need to keep committing syscalls.h to the tree, though. And given how often sysproto.h is included, it likely needs to be committed, though we do fine with the vnode stuff getting things built early enough.

bsdimp commented on src/config.lua:
```diff
@@ -19,10 +20,135 @@
+	NOTSTATIC	= 0x00000100,
+	CAPENABLED	= 0x00000200,
+	SYSMUX		= 0x00000400,
+
```
rather than bits, the known flags might be better as a table so they can be looked up more easily and in a less transliterated from awk way :).

bsdimp commented on src/scarg.lua:
```diff
@@ -2,21 +2,178 @@
+}
+
+local function check_abi_changes(arg)
+	for k, v in pairs(config.known_abi_flags) do
```
Changes when? I know this code is copied from makesyscalls, but it's when this ABI changes from the native ABI. A comment above the function would be good.
